### PR TITLE
Fix import handling for Helm Release

### PR DESCRIPTION
**Pull request by viveklak**
### Proposed changes
Adds ability to import releases.

Note - since releases don't themselves record information about the chart used, repository information is omitted from the code generated.

### Related issues (optional)
https://github.com/pulumi/pulumi-kubernetes/issues/1698

### CODE REVIEW

**Comment by viveklak** [comment]
I have manually verified this. Working on an integration test for this.

**Comment by lblackstone** [comment]
I think this looks good, but it will likely have a merge conflict with #1809 

Not sure which would be easier to merge first

**Comment by silphid** [comment]
I tried it out on the release we're trying to import and we get a warning during preview:
```
Diagnostics:
  kubernetes:helm.sh/v3:Release (ingress-nginx-controller):
    warning: inputs to import do not match the existing resource; importing this resource will fail
```
and equivalent error during update:
```
  kubernetes:helm.sh/v3:Release (ingress-nginx-controller):
    error: inputs to import do not match the existing resource
```
Is there anything we should do to make that work?

Our resource looks like this in TS:
```
    new k8s.helm.v3.Release(
      model.name,
      {
        chart: 'ingress-nginx',
        repositoryOpts: {
          repo: 'https://kubernetes.github.io/ingress-nginx',
        },
        values: values,
        version: model.repoVersion,
        namespace: 'nginx-ingress-controller',
        keyring: '~/.gpg/pubring.gpg',
      },
      {
        parent: this,
        import: 'nginx-ingress-controller/release-kl2buwnx',
        provider: cluster.provider,
      }
    );
```

**Comment by viveklak** [comment]
> We tried out that PR with the release we're trying to import and we get a warning during preview:
> 
> ```
> Diagnostics:
>   kubernetes:helm.sh/v3:Release (ingress-nginx-controller):
>     warning: inputs to import do not match the existing resource; importing this resource will fail
> ```
> 
> and equivalent error during the actual import:
> 
> ```
>   kubernetes:helm.sh/v3:Release (ingress-nginx-controller):
>     error: inputs to import do not match the existing resource
> ```
> 
> Is there anything we should do to make that work?
> 
> Our resource looks like this in TS:
> 
> ```
>     new k8s.helm.v3.Release(
>       model.name,
>       {
>         chart: 'ingress-nginx',
>         repositoryOpts: {
>           repo: 'https://kubernetes.github.io/ingress-nginx',
>         },
>         values: values,
>         version: model.repoVersion,
>         namespace: 'nginx-ingress-controller',
>         keyring: '~/.gpg/pubring.gpg',
>       },
>       {
>         parent: this,
>         import: 'nginx-ingress-controller/release-kl2buwnx',
>         provider: cluster.provider,
>       }
>     );
> ```

Thanks for testing this out already! I discovered this and a couple of similar issues in my tests as well. Essentially we need to drop some computed items from the inputs which are being detected as conflicts. In addition, we set certain fields as defaults which are conflicting with the import. I will have fixes for both in the PR shortly.

File: provider/pkg/provider/provider.go [hunk 1]
```diff
@@ -647,21 +647,19 @@ func (k *kubeProvider) Configure(_ context.Context, req *pulumirpc.ConfigureRequ
 	// Attempt to load the configuration from the provided kubeconfig. If this fails, mark the cluster as unreachable.
 	if !k.clusterUnreachable {
 		config, err := kubeconfig.ClientConfig()
-
-		if kubeClientSettings.Burst != nil {
```
Comment (viveklak) [inline provider/pkg/provider/provider.go:651]: I ran into some panics here in tests which this change fixed.

File: tests/sdk/go/helm-release-import/step1/nginx/.helmignore [hunk 2]
```diff
@@ -0,0 +1,21 @@
+# Patterns to ignore when building packages.
```
Comment (viveklak) [inline tests/sdk/go/helm-release-import/step1/nginx/.helmignore:1]: This is the same as the chart in the local chart example. Its a lot easier to refer to a local chart with the manual chart install and figured its best to copy it. Let me know if we should consolidate instead.

**Comment by viveklak** [comment]
@lblackstone I think this is ready for another look now.

File: tests/sdk/go/go_test.go [hunk 3]
```diff
@@ -18,6 +18,8 @@ import (
 	b64 "encoding/base64"
 	"encoding/json"
 	"fmt"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/contract"
```
Comment (lblackstone) [inline tests/sdk/go/go_test.go:21]: nit: import sorting

File: tests/sdk/go/go_test.go [hunk 4]
```diff
@@ -107,6 +109,22 @@ func TestGo(t *testing.T) {
 		integration.ProgramTest(t, &options)
 	})
 
+	t.Run("Helm Import", func(t *testing.T) {
+		baseDir := filepath.Join(cwd, "helm-release-import", "step1")
+		namespace := getRandomNamespace("importtest")
+		require.NoError(t, createRelease("mynginx", namespace, baseDir, true))
+		defer func() {
+			contract.AssertNoError(deleteRelease("mynginx", namespace))
```
Comment (lblackstone) [inline tests/sdk/go/go_test.go:117]: Should this be `IgnoreError` instead?

File: tests/sdk/go/helm-release-import/step1/nginx/.helmignore (hunk 2 above)
Comment (lblackstone) [reply tests/sdk/go/helm-release-import/step1/nginx/.helmignore:1]: Fine by me

File: tests/sdk/go/manual.go [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package test
+
+import (
+	"fmt"
+	"github.com/pulumi/pulumi-kubernetes/provider/v3/pkg/provider"
```
Comment (lblackstone) [inline tests/sdk/go/manual.go:19]: nit: import sorting

File: tests/sdk/go/manual.go [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+	if err != nil {
+		return err
+	}
+	fmt.Println("Successfully installed release: ", rel.Name)
+	return nil
+}
+
+func deleteRelease(releaseName, releaseNamespace string) error {
+	actionConfig := new(action.Configuration)
+	if err := actionConfig.Init(kube.GetConfig("", "", releaseNamespace), releaseNamespace, os.Getenv("HELM_DRIVER"), func(format string, v ...interface{}) {
```
Comment (lblackstone) [inline tests/sdk/go/manual.go:82]: Do you not need a kubeconfig for the delete?

Comment (viveklak) [reply tests/sdk/go/manual.go]: Yes! Thanks for catching. Fixed.

File: tests/sdk/go/manual.go (hunk 5 above)
Comment (viveklak) [reply tests/sdk/go/manual.go]: Fixed.

File: tests/sdk/go/go_test.go (hunk 4 above)
Comment (viveklak) [reply tests/sdk/go/go_test.go]: Done

File: tests/sdk/go/go_test.go (hunk 3 above)
Comment (viveklak) [reply tests/sdk/go/go_test.go]: Done

**Review by lblackstone — APPROVED** [review]

**Comment by silphid** [comment]
(Sorry, that was me above, just logged as our bot account! 😉)

**Comment by silphid** [comment]
Another input that seems to be computed dynamically and which should maybe be dropped is `keyring`. On my machine, it computes to `/Users/mathieu/.gnupg/pubring.gpg`, which obviously won't be the same on my colleague's machine (and thus always require an update). What do you think? 🤔

**Comment by viveklak** [comment]
@silphid thanks for that. You might want to disable the following from the resource during the import. 
```repositoryOpts: {
          repo: 'https://kubernetes.github.io/ingress-nginx',
},```
You can also drop `keyring: '~/.gpg/pubring.gpg',`.

**Pull request merged**